src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs

```diff
@@ -125,6 +128,13 @@ public void AddDirty(IVisual visual)
                 if (m.HasValue)
                 {
                     var bounds = new Rect(visual.Bounds.Size).TransformToAABB(m.Value);
+
+                    if (_lastPaintBounds != default)
```

Comment:
this is not optimal as last Paint is always the full width/height of the window, so essentially we have full frame invalidation every frame.
Essentially we need to invalidate previous bounds (controls) for the bounds (control) that are changed from previous render. So moving progressbar indicator for example invalidates every frame old/moved from and new/moved to bounds